### PR TITLE
types: handle multiple whitespace in ParseDateFormat

### PR DESCRIPTION
**Pull request by tiancaiamao**
The following date format valid, `ParseDateFormat` should handle it.
```
2018-02-07      00:00:00
``` 

Fix issue https://github.com/pingcap/tidb/issues/5819

### CODE REVIEW

File: types/time_test.go [hunk 1]
```diff
@@ -824,15 +824,19 @@ func (s *testTimeSuite) TestParseDateFormat(c *C) {
...
-		{"2011-11-11 10:10:10.123456", []string{"2011", "11", "11", "10", "10", "10", "123456"}},
-		{"  2011-11-11 10:10:10.123456  ", []string{"2011", "11", "11", "10", "10", "10", "123456"}},
+		{"2012-12-31   11:30:45", []string{"2012", "12", "31", "11", "30", "45"}},
+		{"2018a01b02", nil},
+		{"2018a01b02  00:00:00", nil},
+		{"2011-11-11 10:10:10", []string{"2011", "11", "11", "10", "10", "10"}},
+		{"  2011-11-11 10:10:10  ", []string{"2011", "11", "11", "10", "10", "10"}},
 		{"2011-11-11 10", []string{"2011", "11", "11", "10"}},
-		{"2011-11-11T10:10:10.123456", []string{"2011", "11", "11", "10", "10", "10", "123456"}},
-		{"2011:11:11T10:10:10.123456", []string{"2011", "11", "11", "10", "10", "10", "123456"}},
```
Comment (jackysp) [inline types/time_test.go]: Why remove these tests?

Comment (jackysp) [reply types/time_test.go]: ok

**Comment by tiancaiamao** [comment]
PTAL @jackysp

**Comment by jackysp** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @winkyao

File: types/time.go [hunk 2]
```diff
@@ -528,36 +528,78 @@ func TimestampDiff(unit string, t1 Time, t2 Time) int64 {
...
 
-			seps = append(seps, format[start:i])
-			start = i + 1
+func (h *parseFormatHelper) parseSeperate() {
+	if !h.valid() {
+		return
+	}
+	// Separator is a single none-number char or whitespaces.
+	switch h.format[h.idx] {
+	case ':', '-', '^', '+', '/', '@', '~', '.':
```
Comment (winkyao) [inline types/time.go]: what about '(' and so on? you can exec below in MySQL:

```
mysql> select date('2011-11-11(10:10:10');
+-----------------------------+
| date('2011-11-11(10:10:10') |
+-----------------------------+
| 2011-11-11                  |
+-----------------------------+
1 row in set (0.00 sec)
```

File: types/time.go [hunk 3]
```diff
@@ -528,36 +528,78 @@ func TimestampDiff(unit string, t1 Time, t2 Time) int64 {
...
+		h.idx++
+	case 'T':
```
Comment (winkyao) [inline types/time.go]: Does 'T' has special meaning?

Comment (tiancaiamao) [reply types/time.go]: T occurs as most one time:
2018/01/08T00:00:00   this is ok
2018/01T08T00:00:00  this is bad

File: types/time.go (hunk 2 above)
Comment (tiancaiamao) [reply types/time.go]: Now I begin to doubt whether it worth the  change...

**Comment by shenli** [comment]
Any update?

**Comment by tiancaiamao** [comment]
I think this change is not worthwhile. It cope with multiple whitespace case, but still can't handle many other case to be 100% compatible.

**Pull request closed without merging**